### PR TITLE
improve IIFE docs

### PR DESCRIPTION
**Pull request by kirkwaiblinger**
### Description

Added examples of IIFE use cases, and made changes to confusing and misleading information.

### Motivation

I was pointing a github user to the docs for IIFEs, since they were in a position where they were absolutely required to use multiple statements in a position that only accepted one, which is perfect use case for an IIFE, see https://github.com/marilari88/zod-playground/issues/130#issuecomment-2552222083 (and also https://github.com/marilari88/zod-playground/issues/81). But when I looked at the docs page I had linked... I found it to be extremely difficult to understand, and didn't really address the use case I was referring to at all.

### Related issues and pull requests

### CODE REVIEW

File: files/en-us/glossary/iife/index.md [hunk 1]
```diff
@@ -6,45 +6,74 @@ page-type: glossary-definition
...
+  } else {
+    return "tails";
+  }
+}());
 ```
 
 It is a design pattern which is also known as a {{glossary("Self-Executing Anonymous Function")}} and contains two major parts:
 
-1. The first is the anonymous function with lexical scope enclosed within the [grouping operator](/en-US/docs/Web/JavaScript/Reference/Operators/Grouping) `()`. This prevents accessing variables within the IIFE idiom as well as polluting the global scope.
-2. The second part creates the immediately invoked function expression `()` through which the JavaScript engine will directly interpret the function.
```
Comment (kirkwaiblinger) [inline files/en-us/glossary/iife/index.md:29]: this was particularly incomprehensible to me

File: files/en-us/glossary/iife/index.md [hunk 2]
```diff
@@ -6,45 +6,74 @@ page-type: glossary-definition
 
 {{GlossarySidebar}}
 
-An **IIFE** (Immediately Invoked Function Expression) is a {{glossary("JavaScript")}} {{glossary("function")}} that runs as soon as it is defined.
```
Comment (kirkwaiblinger) [inline files/en-us/glossary/iife/index.md:9]: At first read, this sounds to me as a reader like 
```js
function implicitlyRunsImmediatelyForSomeReason() {};
```
rather than an explicit pattern.

File: files/en-us/glossary/iife/index.md [hunk 3]
```diff
@@ -6,45 +6,74 @@ page-type: glossary-definition
...
 ## Use cases
 
-### Avoid polluting the global namespace
+### Avoid polluting the global namespace in script code
 
 Because our application could include many functions and global variables from different source files, it's
-important to limit the number of global variables. If we have some initiation code that we don't need to use
-again, we could use the IIFE pattern. As we will not reuse the code again, using IIFE in this case is better than
+important to limit the number of variables at the top level of a script
+(this advice [does not apply to module code](/en-US/docs/Web/JavaScript/Guide/Modules#other_differences_between_modules_and_classic_scripts)).
```
Comment (60x20) [inline files/en-us/glossary/iife/index.md:61]: I don't think leaving out modules is a good idea, since it implies that it's okay to pollute the module scope (and `pollute` implies that the scope will keep variables that are no longer needed).
```
const year2000 = new Date('2000');
const currentDate = new Date();
export const isAfterYear2000 = currentDate > year2000;
```
```
export const isAfterYear2000 = (() => {
  const year2000 = new Date('2000');
  const currentDate = new Date();
  return currentDate > year2000;
})();
```

Comment (kirkwaiblinger) [reply files/en-us/glossary/iife/index.md:61]: There's a critical conceptual difference between "I want to avoid an intermediate variable for all the reasons you might ordinarily want to avoid an intermediate variable" and "if I create an intermediate variable here it will be a global variable" (which applies to top level of scripts only). My goal is to convey that difference.

Do you have a suggestion how I might communicate that better? I.e. to communicate that if it's not the top level of a script, this is still a useful pattern, but it doesn't solve nearly as catastrophic a problem as the script global variable problem.

Comment (60x20) [reply files/en-us/glossary/iife/index.md:61]: It doesn't look necessary to mention modules when the focus is already on the global namespace. Though it's true the advice is more lenient with modules.

**Review by 60x20 — COMMENTED** [review]
Looks good without the exclusion of modules.

**Comment by Josh-Cena** [comment]
This page is getting too long. In general glossary entries should try to be a few paragraphs. I will try to move this to the JS reference instead.

**Comment by Josh-Cena** [comment]
Moved the examples around. @kirkwaiblinger @60x20 WDYT?

File: files/en-us/web/javascript/reference/operators/function/index.md [hunk 4]
```diff
@@ -127,18 +127,174 @@ button.addEventListener("click", function (event) {
...
+        if (balance >= amount) {
+          balance -= amount;
+          return balance;
+        }
+        return "Insufficient money";
+      },
+    };
+  })(balance);
+
+const firstAccount = makeWithdraw(100); // "I will do bad things with your money"
```
Comment (kirkwaiblinger) [inline files/en-us/web/javascript/reference/operators/function/index.md:198]: neither here nor there really, but I really dislike this example 
- The specific code example is hard to follow and just.... weird 😆 
- I'm not sure this use pattern in general is that useful especially with `#private` class members these days. But I could be wrong about this

**Comment by kirkwaiblinger** [comment]
@Josh-Cena I'm happy!

File: files/en-us/web/javascript/reference/operators/function/index.md (hunk 4 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/operators/function/index.md:198]: It is a bit weird, but this is exactly how modules work, so I think it's still useful, at least not any less than the `for` loop scoping example below. We already said that "Many traditional use cases of IIFEs have been obsoleted by new syntax features such as modules and block-scoped declarations", so I'd like to keep this for now.

**Review by Josh-Cena — APPROVED** [review]
:+1: Thanks a lot!
